### ISSUE TITLE
The compiler doesn't suggest borrowing `dbg!(x)`

### ISSUE THREAD

**Issue by amab8901** [rust-lang/rust#121795]
```
    let x = "Hello".to_string();
    dbg!(x);
    println!("{x}");
``` 

The above code returns this following error:
```
borrow of moved value: `x`
value borrowed here after move
``` 

To solve this, you'd normally have to add a reference character like this:
```
dbg!(&x);
```
But I think that Rust compiler should be able to understand what i'm trying to do without me having to manually add a reference on the variable `x`. I think this could be achieved by making an implementation that follows a similar logic as the following pseudocode:
```
fn dbg_reference_checker<T>(
    x: T
) -> Result<(), DbgError> {
    if !error_w_moved_value {
        use_moved_value(x)?;
    } else if !error_with_reference {
        use_reference(x)?;
    } else {
        return Err(DbgError::CannotDebugTheVariable)
    }
}

```

**Comment by clubby789** [comment]
`dbg` is just a [regular macro](https://doc.rust-lang.org/src/std/macros.rs.html#352-374), it can't do the high level analysis that would be needed to know if it needed to insert a borrow.

It's specifically designed to move the argument, so you can do, e.g.

```
some_function(dbg!(argument));
```

**Comment by Nilstrieb** [comment]
The error message I'm getting is really silly though
```rust
error[E0382]: use of moved value: `s`
 --> src/main.rs:2:27
  |
2 | let s = vec![0]; dbg!(s); s;
  |     -            -------  ^ value used here after move
  |     |            |
  |     |            value moved here
  |     move occurs because `s` has type `Vec<i32>`, which does not implement the `Copy` trait
  |
help: borrow this binding in the pattern to avoid moving the value
 --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/macros.rs:364:13
  |
36|             ref tmp => {
  |             +++

For more information about this error, try `rustc --explain E0382`.```

So let's repurpose this issue, whose feature request will not be implemented as clubby showed, into a diagnostics issue where the compiler should at least suggest borrowing. I feel like I've seen that before, but not in my test.

**Comment by fmease** [comment]
> changed the title **~~Let `dbg!(x)` automatically reference the variable `x` to prevent error~~ The compiler doesn't suggest borrowing `dbg!(x)`**

In this case closing as duplicate of #120327 which may be fixed by #120990.
